### PR TITLE
stalwart-mail: init at 0.3.1

### PR DESCRIPTION
**Pull request by happysalada**
###### Description of changes

draft of adding the mailservers to nixos.
The reason is draft is because of a compilation failure for jmap (problem with a symlink).
I'm opening this early so people know this is ongoing.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 234672` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>imap-server</li>
    <li>jmap-server</li>
    <li>smtp-server</li>
  </ul>
</details>

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 234672` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>imap-server</li>
    <li>jmap-server</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>smtp-server</li>
  </ul>
</details>

**Comment by teto** [comment]
finally a jmap server ! thanks for looking at this <3

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 234672` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>imap-server</li>
    <li>jmap-server</li>
    <li>smtp-server</li>
  </ul>
</details>

**Comment by happysalada** [comment]
@teto nice to see you here!
I think this is ready for review, let me know if anything.
oh, and if you get to make a nixos module before I do, feel free to tag me in :-)

File: pkgs/servers/mail/stalwart/imap/default.nix
Comment (evanrichter) [inline pkgs/servers/mail/stalwart/imap/default.nix:42]: darwin also needs `SystemConfiguration`

File: pkgs/servers/mail/stalwart/jmap/default.nix
Comment (evanrichter) [inline pkgs/servers/mail/stalwart/jmap/default.nix:40]:
```suggestion
  buildInputs = lib.optionals stdenv.isDarwin [ SystemConfiguration ];

```

also here

**Review by evanrichter — CHANGES_REQUESTED** [review]
testing on aarch64-darwin, just need to add another apple framework

**Comment by evanrichter** [comment]
I also think the package names are too generic and should probably have the name "stalwart" in it.

for example:

- stalwart-imap
- stalwart-jmap
- stalwart-smtp

**Comment by happysalada** [comment]
your comments make sense, I was really thinking the naming should be changed.
After talking with the creator in the upstream repo, there is also a cli package and an install package that most likely need to be packaged again.

All your comments make sense, I should have time soonish to take care of those. Thanks a lot for coming by!

**Comment by mdecimus** [comment]
Hi, thanks for working on this!

For the sake of simplicity I propose to release just one package `stalwart-mail` containing the all-in-one version rather than multiple packages for each component. Although the all-in-one binary is slightly larger, it can also be used as a standalone JMAP, IMAP or SMTP server.

Also, there `stalwart-mail` binary can be compiled to use either a SQLite or FoundationDB backend. However, since the FoundationDB version compiled in the binary needs to match the version of the FDB cluster, I think it is better to only distribute the SQLite version.

**Comment by happysalada** [comment]
I updated by packaging only "mail-server" from stalwart. I called it "stalwart-mail-server". 

I didn't specify any "--package" flag to the build step, letting it build everything. I think we need at least "main", "cli" and "install".

the "nix-init" program detected a dependency on zstd and bzip2, so I left those in as is.

I didn't make the feature set customizable, leaving it for later potentially.

I didn't respect the pattern of passing darwin frameworks from all-packages. (I still don't understand why it should be done that way).

I just outlined areas where people might disagree/comment, let me know if anything.

File: pkgs/servers/mail/stalwart/default.nix
Comment (gkleen) [inline pkgs/servers/mail/stalwart/default.nix:64]:
```suggestion
    description = "Secure & Modern All-in-One Mail Server (IMAP, JMAP, SMTP)";
```

**Comment by happysalada** [comment]
ofborg is getting two tests failure
```
---- core::if_block::tests::eval_if stdout ----
thread 'core::if_block::tests::eval_if' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', crates/smtp/src/core/if_block.rs:248:62
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
---- outbound::dane::verify::test::dane_test stdout ----
thread 'outbound::dane::verify::test::dane_test' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', crates/smtp/src/outbound/dane/verify.rs:223:53
```
one of the failures seems to originate from https://github.com/stalwartlabs/mail-server/blob/main/crates/smtp/src/outbound/dane/verify.rs#L223
Do the tests expect a certain file structure ? Should I run something in the pre test phase to make sure the expected file structure is there ?
@mdecimus

**Comment by happysalada** [comment]
after disabling those tests, some more tests failed, I traced those back to this method
https://github.com/bluejekyll/trust-dns/blob/bf04bc8d3efec41d29cf558416b3dbed2d26ae02/crates/resolver/src/system_conf/unix.rs#L29
The tests, expect something in `/etc/resolv.conf`, and not finding anything fails just about half of the tests.

I've disabled the tests for now, but someone with more networking knowledge should know how to enable those again.

**Comment by happysalada** [comment]
ofborg passes, this should be good to merge unless someone has additional comments.
Not in a rush to merge, so just let me know if you need me to wait until you have time to review. Otherwise thinking about merging in case no one comments for 3-4 days.

**Comment by sinavir** [comment]
Hello, very nice work. I have tested a bit and it seems to me that not all tests are failing.

Don't you think we should only disable the failing tests ?

**Comment by sinavir** [comment]
Also @happysalada have you already worked on a module ?

**Comment by happysalada** [comment]
@sinavir thanks for coming by !

Around disabling tests.
I build the package originally and got 2 failures in the tests for smtp. Then after disabling those, i got 17/34 failures for smtp again, all related to the missing resolve.conf .
I dont have access to a powerful machine at the moment, so i gave up after that. If you have time and are willing to find which tests need to be disabled, id be happy to add them to this PR. (I can make a commit based on a suggestion and squash them to retain attribution)

**Comment by happysalada** [comment]
regarding the module, I haven't started, Mauro was nice enough to point us to a pre-existing systemd service file at https://github.com/stalwartlabs/mail-server/blob/main/resources/systemd/stalwart-mail.service

I'd be happy to review and test a PR if you wanted to make the module contribution.
I haven't looked at how secrets are handled, but it's usually a problem with most binaries that take a secret as an environment variable or as plain text. The ideal path for supporting secrets with nixos goes through LoadCredentials , but it usually requires upstreaming reading secrets from path. (I don't know how stalwart handles secrets, it might be good from the get go, just sharing a potential pitfal).

**Comment by mdecimus** [comment]
@happysalada  Stalwart loads three types of secrets when starting:

- TLS certificates: These are configured under the `certificate` section of the configuration file. Certificates can be embedded in the configuration file or read from an external file.
- DKIM keys: These are configured under the `signature` section of the configuration file and, just like TLS certs, they can also be embedded or read from a configuration file.
- OAuth encryption key: It is the key used to encrypt the OAuth tokens and is configured in the `oauth.key` attribute of the configuration file.

In a standard setup, both the configuration file and all external certificate/key files have a `600` chmod and are owned by `root` or the stalwart service account. In Linux, Stalwart is usually started as root in order to bind to the privileged ports and then the permissions are dropped to the stalwart user specified in the configuration file.

**Comment by evanrichter** [comment]
Loading secrets from files is ideal for agenix, and similar solutions. It's best practice to not have any secrets directly in config files because nix will likely put them in the world-readable /nix/store before activation. So I'm hoping the OAuth secret can also be configured as a file path as well!

**Comment by evanrichter** [comment]
> I called it "stalwart-mail-server".

not to bike-shed too much, but I think "-server" is redundant in this case. Less than 10% of packages in the pkgs/server/ folder have "server" in the package name, and many times it's because that's part of the upstream project name. It's simpler to just use the package's main binary name as the package name in my (relatively new) experience

**Comment by happysalada** [comment]
switched to stalwart-mail.

I won't have time to work on the nixos-module for the next month probably. So if anyone wants to pick it up they are welcome. I should be back home in september.

**Comment by mdecimus** [comment]
Quick update: Version `0.3.2` now allows reading OAuth keys from files as well. By default the configuration script generates a random OAuth key which is embedded in the configuration file (`oauth.key` attribute). In Nix you could either extract the auto-generated key to a file or simply ignore it and replace with the path to a key you generated.

**Comment by happysalada** [comment]
thanks a lot for the quick turnover!
I've made the update in https://github.com/NixOS/nixpkgs/pull/245842
(I've also added an automated update script that should make updates automaticalyl with 2-3 days of delay).

**Comment by pacien** [comment]
FYI, module here: https://github.com/NixOS/nixpkgs/pull/253022

**Pull request merged**
